temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowInternal.java

```diff
@@ -199,6 +188,73 @@ public static void registerListener(Object implementation) {
           .registerSignalHandlers(
               new WorkflowOutboundCallsInterceptor.RegisterSignalHandlersInput(requests));
     }
+
+    // Get all validators and lazily assign them to update handlers as we see them.
+    Map<String, POJOWorkflowMethodMetadata> validators = new HashMap<>();
+    for (POJOWorkflowMethodMetadata methodMetadata : workflowMetadata.getUpdateValidatorMethods()) {
+      Method method = methodMetadata.getWorkflowMethod();
+      UpdateValidateMethod updateValidatorMethod = method.getAnnotation(UpdateValidateMethod.class);
+      if (validators.containsKey(updateValidatorMethod.updateName())) {
+        throw new IllegalArgumentException(
+            "Duplicate validator for update handle " + updateValidatorMethod.updateName());
+      }
+      validators.put(updateValidatorMethod.updateName(), methodMetadata);
+    }
+
+    List<WorkflowOutboundCallsInterceptor.UpdateRegistrationRequest> updateRequests =
+        new ArrayList<>();
+    for (POJOWorkflowMethodMetadata methodMetadata : workflowMetadata.getUpdateMethods()) {
+      Method method = methodMetadata.getWorkflowMethod();
+      UpdateMethod updateMethod = method.getAnnotation(UpdateMethod.class);
+      String updateMethodName = updateMethod.name();
+      if (updateMethodName.isEmpty()) {
+        updateMethodName = method.getName();
+      }
+      // Check if any validators claim they are the validator for this update
+      POJOWorkflowMethodMetadata validatorMethodMetadata = validators.get(updateMethodName);
```

Comment:
```suggestion
      POJOWorkflowMethodMetadata validatorMethodMetadata = validators.remove(updateMethodName);
```

No need for a separate remove call below